application-numbered-content-reference/application-numbered-content-reference-macro/src/main/java/org/xwiki/contrib/numbered/content/reference/internal/ReferenceMacro.java

```diff
@@ -159,7 +167,26 @@ private String figuresNumber(String id, Block rootBlock)
 
     private boolean hasId(FigureBlock figureBlock, String id)
     {
-        return figureBlock.<IdBlock>getBlocks(new ClassBlockMatcher(IdBlock.class), Block.Axes.DESCENDANT).stream()
-            .anyMatch(it -> Objects.equals(it.getName(), id));
+        return figureBlock.getFirstBlock(block -> {
+            String idParameter = block.getParameter(ID_PARAMETER_KEY);
+            if (idParameter != null) {
+                // Match the id parameter of any block.
+                return Objects.equals(idParameter, id);
+            }
+
+            if (block instanceof IdBlock) {
+                // Match the id of the id block.
+                IdBlock idBlock = (IdBlock) block;
+                return Objects.equals(idBlock.getName(), id);
+            }
+
+            if (block instanceof ImageBlock) {
+                // Match the id of the image block.
+                ImageBlock imageBlock = (ImageBlock) block;
+                return Objects.equals(imageBlock.getId(), id);
+            }
```

Comment:
Note 1: If we ever have to add on if in this code, we'll need to change the design to make it more modular :)
Note 2: I don't think we have a formal rule regarding the number of return in a method, but I tend to prefer a single one at the end of the method.

Reply:
We do have a formal rule, and it says five returns are allowed: https://github.com/xwiki/xwiki-commons/blob/master/xwiki-commons-tools/xwiki-commons-tool-verification-resources/src/main/resources/checkstyle.xml#L326 But I'll change it.